File: web/test/generate-fixtures.ts
/**
 * Generate test fixtures from TypeScript for Python cross-compatibility testing.
 *
 * This script generates JSON fixture files that can be used to verify
 * that Python implementations produce identical results to TypeScript.
 *
 * Usage:
 *   npm run generate-fixtures
 */

import { writeFileSync, mkdirSync, existsSync } from 'fs';
import { join } from 'path';
import { webcrypto } from 'crypto';

// Polyfill crypto for Node.js environment
if (typeof globalThis.crypto === 'undefined') {
  // @ts-ignore
  globalThis.crypto = webcrypto;
}

import { FixedDistributionClient } from '../src/fixed-client.ts';
import {
  deriveKey,
  encryptPayload,
  bytesToBase64,
  randomBytes,
} from '../src/crypto.ts';
import { encryptSecretBlob, generateSecret } from '../src/secret.ts';
import {
  huffmanEncode,
  compressPayload,
  DEFAULT_FREQUENCIES,
} from '../src/huffman.ts';
import { encodeMessage } from '../src/stego.ts';
import {
  PBKDF2_ITERATIONS,
  PAYLOAD_KEY_SIZE,
  SECRET_VERSION,
  COMPRESSION_HUFFMAN,
  COMPRESSION_NONE,
} from '../src/types.ts';

const FIXTURES_DIR = join(__dirname, '../../tests/fixtures');

// Ensure fixtures directory exists
if (!existsSync(FIXTURES_DIR)) {
  mkdirSync(FIXTURES_DIR, { recursive: true });
}

async function generateCryptoVectors(): Promise<void> {
  console.log('Generating crypto vectors from TypeScript...');

  const pbkdf2Vectors: Array<{
    password: string;
    salt: string;
    expected_key: string;
  }> = [];

  // Test various password/salt combinations
  const testCases = [
    { password: 'test123', salt: '0123456789abcdef' },
    { password: 'password', salt: 'saltsaltsaltsalt' },
    { password: '', salt: 'emptypwd00000000' },
    { password: 'long_password_with_special_chars!@#$%', salt: 'anothersalt12345' },
    { password: '世界', salt: 'unicodepw1234567' },
  ];

  for (const tc of testCases) {
    const salt = new TextEncoder().encode(tc.salt);
    const key = await deriveKey(tc.password, salt);
    const keyBytes = await crypto.subtle.exportKey('raw', key);

    pbkdf2Vectors.push({
      password: tc.password,
      salt: bytesToBase64(salt),
      expected_key: bytesToBase64(new Uint8Array(keyBytes)),
    });
  }

  // Generate AES-GCM vectors
  const aesgcmVectors: Array<{
    key: string;
    plaintext: string;
    ciphertext: string;
  }> = [];

  const testPlaintexts = [
    new TextEncoder().encode('Hello, World!'),
    new Uint8Array(0), // Empty
    new Uint8Array([0x00, 0xff, 0x00, 0xff, 0x00, 0xff, 0x00, 0xff, 0x00, 0xff, 0x00, 0xff, 0x00, 0xff, 0x00, 0xff, 0x00, 0xff, 0x00, 0xff]), // Binary
    new TextEncoder().encode('Unicode: 世界'),
    new Uint8Array(1000).fill(65), // Larger payload
  ];

  for (const plaintext of testPlaintexts) {
    // Use a fixed key for reproducibility
    const key = new TextEncoder().encode('0123456789abcdef0123456789abcdef');
    const ciphertext = await encryptPayload(plaintext, key);

    aesgcmVectors.push({
      key: bytesToBase64(key),
      plaintext: bytesToBase64(plaintext),
      ciphertext: bytesToBase64(ciphertext),
    });
  }

  const cryptoVectors = {
    pbkdf2_vectors: pbkdf2Vectors,
    aesgcm_vectors: aesgcmVectors,
    generated_by: 'typescript',
  };

  writeFileSync(
    join(FIXTURES_DIR, 'crypto_vectors_ts.json'),
    JSON.stringify(cryptoVectors, null, 2)
  );
  console.log(`  -> ${join(FIXTURES_DIR, 'crypto_vectors_ts.json')}`);
}

async function generateSecretBlobs(): Promise<void> {
  console.log('Generating secret blobs from TypeScript...');

  const blobs: Array<{
    encrypted: string;
    password: string;
    expected: {
      version: number;
      k: number;
      knock: number[];
      preamble_tokens: number;
      suffix_tokens: number;
      temperature: number;
      payload_key_length: number;
    };
  }> = [];

  const testConfigs = [
    {
      k: 16,
      knock: [0, 1, 2, 3, 4, 5],
      password: 'ts_test_password_123',
      preambleTokens: 10,
      suffixTokens: 10,
      temperature: 0.8,
    },
    {
      k: 8,
      knock: [0, 1, 2, 3],
      password: 'short',
      preambleTokens: 5,
      suffixTokens: 5,
      temperature: 0.5,
    },
  ];

  for (const config of testConfigs) {
    const secret = generateSecret({
      k: config.k,
      knock: config.knock,
      preambleTokens: config.preambleTokens,
      suffixTokens: config.suffixTokens,
      temperature: config.temperature,
    });

    const encrypted = await encryptSecretBlob(secret, config.password);

    blobs.push({
      encrypted,
      password: config.password,
      expected: {
        version: secret.version,
        k: secret.k,
        knock: secret.knock,
        preamble_tokens: secret.preamble_tokens,
        suffix_tokens: secret.suffix_tokens,
        temperature: secret.temperature,
        payload_key_length: secret.payload_key.length,
      },
    });
  }

  const secretBlobs = {
    blobs,
    generated_by: 'typescript',
  };

  writeFileSync(
    join(FIXTURES_DIR, 'secret_blobs_ts.json'),
    JSON.stringify(secretBlobs, null, 2)
  );
  console.log(`  -> ${join(FIXTURES_DIR, 'secret_blobs_ts.json')}`);
}

async function generateHuffmanData(): Promise<void> {
  console.log('Generating Huffman data from TypeScript...');

  const compressed: Array<{
    original: string;
    compressed: string;
    compression_type: number;
  }> = [];

  const testData = [
    new TextEncoder().encode('The quick brown fox jumps over the lazy dog.'),
    new TextEncoder().encode('Hello World'.repeat(50)),
    new Uint8Array([0x00, 0xff].concat(Array(18).fill(0x00).concat(Array(18).fill(0xff)))),
    new TextEncoder().encode('Unicode: 世界'),
    new TextEncoder().encode('AAAAAAAAAA'),
    new Uint8Array(0),
    new Uint8Array(Array.from({ length: 95 }, (_, i) => i + 32)), // Printable ASCII
  ];

  for (const data of testData) {
    const [compData, compType] = compressPayload(data, DEFAULT_FREQUENCIES);

    compressed.push({
      original: bytesToBase64(data),
      compressed: bytesToBase64(compData),
      compression_type: compType,
    });
  }

  const huffmanData = {
    compressed,
    default_frequencies: DEFAULT_FREQUENCIES,
    generated_by: 'typescript',
  };

  writeFileSync(
    join(FIXTURES_DIR, 'huffman_data_ts.json'),
    JSON.stringify(huffmanData, null, 2)
  );
  console.log(`  -> ${join(FIXTURES_DIR, 'huffman_data_ts.json')}`);
}

async function generateStegoRoundtrip(): Promise<void> {
  console.log('Generating stego roundtrip data from TypeScript...');

  const client = new FixedDistributionClient(32);

  const encoded: Array<{
    secret_blob: string;
    password: string;
    prompt: string;
    cover_text: string;
    expected_message: string;
  }> = [];

  const testCases = [
    {
      message: new TextEncoder().encode('Secret message from TypeScript'),
      prompt: 'Once upon a time',
      password: 'ts_stego_test_123',
      k: 16,
      knock: [0, 1, 2, 3, 4, 5],
    },
    {
      message: new Uint8Array(0), // Empty message
      prompt: 'Test: ',
      password: 'ts_empty_test',
      k: 16,
      knock: [0, 1, 2, 3, 4, 5],
    },
  ];

  for (const tc of testCases) {
    const secret = generateSecret({
      k: tc.k,
      knock: tc.knock,
      preambleTokens: 5,
      suffixTokens: 5,
      temperature: 0.8,
    });

    const secretBlob = await encryptSecretBlob(secret, tc.password);
    const coverText = await encodeMessage(
      tc.message,
      secret,
      client,
      tc.prompt,
      true
    );

    encoded.push({
      secret_blob: secretBlob,
      password: tc.password,
      prompt: tc.prompt,
      cover_text: coverText,
      expected_message: bytesToBase64(tc.message),
    });
  }

  const stegoRoundtrip = {
    encoded,
    generated_by: 'typescript',
  };

  writeFileSync(
    join(FIXTURES_DIR, 'stego_roundtrip_ts.json'),
    JSON.stringify(stegoRoundtrip, null, 2)
  );
  console.log(`  -> ${join(FIXTURES_DIR, 'stego_roundtrip_ts.json')}`);
}

async function main(): Promise<void> {
  console.log('Generating TypeScript fixtures for cross-compatibility testing\n');

  await generateCryptoVectors();
  await generateSecretBlobs();
  await generateHuffmanData();
  await generateStegoRoundtrip();

  console.log('\nAll TypeScript fixtures generated successfully!');
  console.log(`Fixtures directory: ${FIXTURES_DIR}`);
}

main().catch(console.error);

File: tests/cross_compat/generate_fixtures.py
#!/usr/bin/env python3
"""
Generate test fixtures for cross-compatibility testing.

This script generates JSON fixture files that can be used to verify
that TypeScript implementations produce identical results to Python.

Usage:
    python tests/cross_compat/generate_fixtures.py
"""

import base64
import json
import sys
from pathlib import Path

# Add project root to path
sys.path.insert(0, str(Path(__file__).parent.parent.parent))

from lm_client import FixedDistributionClient
from stego_secret import (
    COMPRESSION_HUFFMAN,
    COMPRESSION_NONE,
    DEFAULT_FREQUENCIES,
    compress_payload,
    derive_key,
    encode_message,
    encrypt_payload,
    encrypt_secret_blob,
    generate_secret,
    huffman_encode,
)

FIXTURES_DIR = Path(__file__).parent.parent / "fixtures"


def generate_crypto_vectors() -> dict:
    """Generate PBKDF2 and AES-GCM test vectors."""
    print("Generating crypto vectors...")

    pbkdf2_vectors = []

    # Test various password/salt combinations
    test_cases = [
        ("test123", b"0123456789abcdef"),
        ("password", b"saltsaltsaltsalt"),
        ("", b"emptypwd00000000"),  # Empty password
        ("long_password_with_special_chars!@#$%", b"anothersalt12345"),
        ("\u4e16\u754c", b"unicodepw1234567"),  # Unicode password
    ]

    for password, salt in test_cases:
        key = derive_key(password, salt)
        pbkdf2_vectors.append(
            {
                "password": password,
                "salt": base64.b64encode(salt).decode(),
                "expected_key": base64.b64encode(key).decode(),
            }
        )

    # Generate AES-GCM vectors
    aesgcm_vectors = []

    test_plaintexts = [
        b"Hello, World!",
        b"",  # Empty
        b"\x00\xff" * 10,  # Binary
        "Unicode: \u4e16\u754c".encode("utf-8"),
        b"A" * 1000,  # Larger payload
    ]

    for plaintext in test_plaintexts:
        # Use a fixed key for reproducibility
        key = b"0123456789abcdef0123456789abcdef"  # 32 bytes
        ciphertext = encrypt_payload(plaintext, key)

        aesgcm_vectors.append(
            {
                "key": base64.b64encode(key).decode(),
                "plaintext": base64.b64encode(plaintext).decode(),
                "ciphertext": base64.b64encode(ciphertext).decode(),
            }
        )

    return {
        "pbkdf2_vectors": pbkdf2_vectors,
        "aesgcm_vectors": aesgcm_vectors,
    }


def generate_secret_blobs() -> dict:
    """Generate encrypted secret blob test data."""
    print("Generating secret blobs...")

    blobs = []

    # Test various secret configurations
    test_configs = [
        {
            "k": 16,
            "knock": [0, 1, 2, 3, 4, 5],
            "password": "test_password_123",
            "preamble_tokens": 10,
            "suffix_tokens": 10,
            "temperature": 0.8,
        },
        {
            "k": 8,
            "knock": [0, 1, 2, 3],
            "password": "short",
            "preamble_tokens": 5,
            "suffix_tokens": 5,
            "temperature": 0.5,
        },
        {
            "k": 32,
            "knock": [0, 5, 10, 15, 20, 25, 30],
            "password": "unicode_\u4e16\u754c",
            "preamble_tokens": 15,
            "suffix_tokens": 15,
            "temperature": 1.0,
        },
    ]

    for config in test_configs:
        secret = generate_secret(
            k=config["k"],
            knock=config["knock"],
            preamble_tokens=config["preamble_tokens"],
            suffix_tokens=config["suffix_tokens"],
            temperature=config["temperature"],
        )

        encrypted = encrypt_secret_blob(secret, config["password"])

        blobs.append(
            {
                "encrypted": encrypted,
                "password": config["password"],
                "expected": {
                    "version": secret["version"],
                    "k": secret["k"],
                    "knock": secret["knock"],
                    "preamble_tokens": secret["preamble_tokens"],
                    "suffix_tokens": secret["suffix_tokens"],
                    "temperature": secret["temperature"],
                    "payload_key_length": len(secret["payload_key"]),
                },
            }
        )

    return {"blobs": blobs}


def generate_huffman_data() -> dict:
    """Generate Huffman compression test data."""
    print("Generating Huffman data...")

    compressed = []

    # Test various data types
    test_data = [
        b"The quick brown fox jumps over the lazy dog.",
        b"Hello World" * 50,  # Repetitive
        b"\x00\xff" * 10,  # Binary
        "Unicode: \u4e16\u754c".encode("utf-8"),
        b"AAAAAAAAAA",  # Very repetitive
        b"",  # Empty
        bytes(range(32, 127)),  # Printable ASCII
    ]

    for data in test_data:
        comp_data, comp_type = compress_payload(data, DEFAULT_FREQUENCIES)

        compressed.append(
            {
                "original": base64.b64encode(data).decode(),
                "compressed": base64.b64encode(comp_data).decode(),
                "compression_type": comp_type,
            }
        )

    # Also include the default frequencies for verification
    return {
        "compressed": compressed,
        "default_frequencies": DEFAULT_FREQUENCIES,
    }


def generate_stego_roundtrip() -> dict:
    """Generate full stego encode/decode test data."""
    print("Generating stego roundtrip data...")

    client = FixedDistributionClient(vocab_size=32)
    encoded = []

    # Test various messages
    test_cases = [
        {
            "message": b"Secret message",
            "prompt": "Once upon a time",
            "password": "stego_test_123",
            "k": 16,
            "knock": [0, 1, 2, 3, 4, 5],
        },
        {
            "message": b"",  # Empty message
            "prompt": "Test: ",
            "password": "empty_test",
            "k": 16,
            "knock": [0, 1, 2, 3, 4, 5],
        },
        {
            "message": "Unicode \u4e16\u754c!".encode("utf-8"),
            "prompt": "Story time: ",
            "password": "unicode_pw",
            "k": 16,
            "knock": [0, 1, 2, 3, 4, 5],
        },
    ]

    for case in test_cases:
        secret = generate_secret(
            k=case["k"],
            knock=case["knock"],
            preamble_tokens=5,
            suffix_tokens=5,
            temperature=0.8,
        )

        # Encrypt the secret
        secret_blob = encrypt_secret_blob(secret, case["password"])

        # Encode the message
        cover_text = encode_message(
            case["message"],
            secret,
            client,
            prompt=case["prompt"],
            compress=True,
        )

        encoded.append(
            {
                "secret_blob": secret_blob,
                "password": case["password"],
                "prompt": case["prompt"],
                "cover_text": cover_text,
                "expected_message": base64.b64encode(case["message"]).decode(),
            }
        )

    return {"encoded": encoded}


def main():
    """Generate all fixture files."""
    FIXTURES_DIR.mkdir(parents=True, exist_ok=True)

    # Generate crypto vectors
    crypto_vectors = generate_crypto_vectors()
    with open(FIXTURES_DIR / "crypto_vectors.json", "w") as f:
        json.dump(crypto_vectors, f, indent=2)
    print(f"  -> {FIXTURES_DIR / 'crypto_vectors.json'}")

    # Generate secret blobs
    secret_blobs = generate_secret_blobs()
    with open(FIXTURES_DIR / "secret_blobs.json", "w") as f:
        json.dump(secret_blobs, f, indent=2)
    print(f"  -> {FIXTURES_DIR / 'secret_blobs.json'}")

    # Generate Huffman data
    huffman_data = generate_huffman_data()
    with open(FIXTURES_DIR / "huffman_data.json", "w") as f:
        json.dump(huffman_data, f, indent=2)
    print(f"  -> {FIXTURES_DIR / 'huffman_data.json'}")

    # Generate stego roundtrip data
    stego_roundtrip = generate_stego_roundtrip()
    with open(FIXTURES_DIR / "stego_roundtrip.json", "w") as f:
        json.dump(stego_roundtrip, f, indent=2)
    print(f"  -> {FIXTURES_DIR / 'stego_roundtrip.json'}")

    print("\nAll fixtures generated successfully!")
    print(f"Fixtures directory: {FIXTURES_DIR}")


if __name__ == "__main__":
    main()
